flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/PartitionedFileWriter.java

```diff
@@ -0,0 +1,211 @@
+	private final long[] subpartitionBytes;
+
+	/** Number of buffers written for each subpartition in the current region. */
+	private final int[] subpartitionBuffers;
+
+	/** Opened data file channel of the target {@link PartitionedFile}. */
+	private FileChannel dataFileChannel;
+
+	/** Opened index file channel of the target {@link PartitionedFile}. */
+	private FileChannel indexFileChannel;
+
+	/** Number of bytes written to the target {@link PartitionedFile}. */
+	private long totalBytesWritten;
+
+	/** Number of regions written to the target {@link PartitionedFile}. */
+	private int numRegions;
+
+	/** Current subpartition to write. Buffer writing must be in subpartition order within each region. */
+	private int currentSubpartition;
+
+	/** Whether all index data can be cached in memory or not. */
+	private boolean canCacheAllIndexData = true;
+
+	/** Whether this file writer is finished. */
+	private boolean isFinished;
+
+	public PartitionedFileWriter(String basePath, int numSubpartitions) {
+		checkArgument(basePath != null, "Base path must not be null.");
+		checkArgument(numSubpartitions > 0, "Illegal number of subpartitions.");
+
+		this.numSubpartitions = numSubpartitions;
+		this.subpartitionBytes = new long[numSubpartitions];
+		this.subpartitionBuffers = new int[numSubpartitions];
+		this.dataFilePath = new File(basePath + PartitionedFile.DATA_FILE_SUFFIX).toPath();
+		this.indexFilePath = new File(basePath + PartitionedFile.INDEX_FILE_SUFFIX).toPath();
+
+		this.indexBuffer = ByteBuffer.allocate(100 * 1024 * PartitionedFile.INDEX_ENTRY_SIZE);
+		indexBuffer.order(PartitionedFile.DEFAULT_BYTE_ORDER);
+	}
+
+	/**
+	 * Opens the {@link PartitionedFile} for writing.
+	 *
+	 * <p>Note: The caller is responsible for releasing the failed {@link PartitionedFile} if any exception
+	 * occurs.
+	 */
+	public void open() throws IOException {
+		checkState(dataFileChannel == null && indexFileChannel == null, "Partitioned file is already opened.");
+
+		dataFileChannel = openFileChannel(dataFilePath);
+		indexFileChannel = openFileChannel(indexFilePath);
+	}
+
+	private FileChannel openFileChannel(Path path) throws IOException {
+		return FileChannel.open(path, StandardOpenOption.CREATE_NEW, StandardOpenOption.WRITE);
+	}
+
+	/**
+	 * Persists the region index of the current data region and starts a new region to write.
+	 *
+	 * <p>Note: The caller is responsible for releasing the failed {@link PartitionedFile} if any exception
+	 * occurs.
+	 */
+	public void startNewRegion() throws IOException {
+		checkState(!isFinished, "File writer is already finished.");
+		checkState(dataFileChannel != null && indexFileChannel != null, "Must open the partitioned file first.");
+
+		writeRegionIndex();
+	}
+
+	private void writeRegionIndex() throws IOException {
+		if (Arrays.stream(subpartitionBytes).sum() > 0) {
+			for (int subpartition = 0; subpartition < numSubpartitions; ++subpartition) {
```

Comment:
I guess the current way assumes the data redistributed in almost all the subpartitions for every flush call. If in the data skew scenario like only 10 of 1000 subpartitions have actual data for some flush, then the index buffer will be written many empty holes to waste space and execution time. 

I am curious that is it by design to have to write all the subpartitions in the same sequence since I have not touched the detail logics on view reader side? Or it is not very performance sensitive for easy implementation ATM?

Reply:
At reading side, we use (subpartition index) × (index entry size) locate an index entry for the given subpartition in a data region. And you are right, it's not very performance sensitive.

Reply:
Ok, let's further drive into it if really necessary future.